testing/listener.go

```diff
@@ -236,6 +236,104 @@ func (tl *TestListener) AwaitData(workflowID string, predicate func(*StateData)
 	tl.AwaitDataFor(workflowID, tl.DefaultWait, predicate)
 }
 
+func (tl *TestListener) ExpectStateFor(workflowID string, waitFor time.Duration, state string, when func()) {
+	proxy := make(chan interface{})
+	go func() {
+		for {
+			proxy <- <-tl.RegisterStateInterest(workflowID)
+		}
+	}()
+	tl.expectFor(workflowID, waitFor, when, proxy, func(i interface{}) bool {
+		return i.(string) == state
+	})
+}
+
+func (tl *TestListener) ExpectState(workflowID, state string, when func()) {
+	tl.ExpectStateFor(workflowID, tl.DefaultWait, state, when)
+}
+
+func (tl *TestListener) ExpectEventFor(workflowID string, waitFor time.Duration, when func(), predicate func(*swf.HistoryEvent) bool) {
+	proxy := make(chan interface{})
+	go func() {
+		for {
+			proxy <- <-tl.RegisterHistoryInterest(workflowID)
+		}
+	}()
+	tl.expectFor(workflowID, waitFor, when, proxy, func(i interface{}) bool {
+		return predicate(i.(*swf.HistoryEvent))
+	})
+}
+
+func (tl *TestListener) ExpectEvent(workflowID string, when func(), predicate func(*swf.HistoryEvent) bool) {
+	tl.ExpectEventFor(workflowID, tl.DefaultWait, when, predicate)
+}
+
+func (tl *TestListener) ExpectDecisionFor(workflowID string, waitFor time.Duration, when func(), predicate func(*swf.Decision) bool) {
+	proxy := make(chan interface{})
+	go func() {
+		for {
+			proxy <- <-tl.RegisterDecisionInterest(workflowID)
+		}
+	}()
+	tl.expectFor(workflowID, waitFor, when, proxy, func(i interface{}) bool {
+		return predicate(i.(*swf.Decision))
+	})
+}
+
+func (tl *TestListener) ExpectDecision(workflowID string, when func(), predicate func(*swf.Decision) bool) {
+	tl.ExpectDecisionFor(workflowID, tl.DefaultWait, when, predicate)
+}
+
+func (tl *TestListener) ExpectDataFor(workflowID string, waitFor time.Duration, when func(), predicate func(*StateData) bool) {
+	proxy := make(chan interface{})
+	go func() {
+		for {
+			proxy <- <-tl.RegisterDataInterest(workflowID)
+		}
+	}()
+	tl.expectFor(workflowID, waitFor, when, proxy, func(i interface{}) bool {
+		return predicate(i.(*StateData))
+	})
+}
+
+func (tl *TestListener) ExpectData(workflowID string, when func(), predicate func(*StateData) bool) {
+	tl.ExpectDataFor(workflowID, tl.DefaultWait, when, predicate)
+}
+
+func (tl *TestListener) expectFor(workflowID string, waitFor time.Duration, when func(), interest chan interface{}, predicate func(interface{}) bool) {
```

Comment:
there's still a lot of duplication here for the different types, but I think I got it as condensed as I can get it without using reflection :-/